### PR TITLE
Proposal? Change fake server to more of a dumb key/value store

### PR DESCRIPTION
**Pull request by mattwynne**
I was thinking about this code today on a long drive with nobody to talk to and a fair bit of coffee in my veins.

I had two insights/puzzled after I read the code last night, which my brain was grinding on during the drive:

- both the client and the server would need to know (and agree on) how to generate an ID for a message, otherwise after you've called `GET /messages?q=...` you wouldn't know what IDs you have to be able to do future comparisons with a `GET /messages-list` result. As a recovering perfectionist, this kind of [connascence](https://www.youtube.com/watch?app=desktop&v=22vYwcfQnk8) bothers my brain.
- although we happen to be using it for messages, this is basically just a key/value store. Maybe it would read better if we acknowledged that? My brain wondered.

So I tried it out in this branch. Take a look. Maybe I missed something about the IDs. Probably the key/value store abstraction isn't all that important in the scheme of the whole project. But I needed to get the idea out of my brain and see what it looked like. And then I thought I might as well share it. 😄

### CODE REVIEW

File: fake-server/src/app.ts [hunk 1]
```diff
@@ -94,28 +92,19 @@ app.get("/messages", async (req, res) => {
...
-app.post("/messages", async (req, res) => {
-  // we do the stringification server-side to ensure that the client can't send
-  // a malformed message
-  // TODO: do we need to worry about this? we could have the client send both the ID and some text, and we just act as a key/value store -- @matt
-  let message = JSON.stringify(req.body)
-  let hash = crypto.createHash("sha256").update(message).digest("hex")
-  await fs.writeFile(`data/${hash}`, JSON.stringify(req.body))
-  res.redirect("/")
+app.post("/docs/:id", async (req, res) => {
+  await fs.writeFile(`data/${req.params.id}`, req.body)
```
[the remark this answers is not in the document]
Comment (mattwynne) [reply fake-server/src/app.ts:106]: I saw your comment above about "we do the stringification server-side to ensure that the client can't send a malformed message" but I wasn't really sure how strongly you were worried about that. It seems to me we can make the server dumber if we trust the clients not to do dumb things, but maybe I'm missing something.

**Comment by blaine** [comment]
Some quick thoughts from my phone (excuse brevity)

Because the IDs are a hash, they're guaranteed to be the same for the same message. 😅

It's a bit of a headfuck, I've been in the content-addressable mines so long that I probably didn't properly explain.

It effectively is a key-value store, but the keys are hashes, so they're not arbitrary. I think we should stick to "message" as the noun, because it keeps us to the correct domain.

Re: the server generating the hash (key), I went back and forth, but ultimately I think it's probably better to have the server do it, because a temporarily buggy/misconfigured/malicious client could cause all sorts of unexpected issues otherwise. The client will still generate the hashes of its messages, but this way they have to be done the same way as the server or it won't work (that's a good thing!)

**Comment by mattwynne** [comment]
OK, my thinking is often about flexibility for future change, so my worry about having both sides having to know how to generate the key was from that angle, (e.g. if we decided to use a shorter hash, as we briefly discussed) but I see the other side of the trade-off now. I had not considered the possibility of a malicious client!

**Comment by mattwynne** [comment]
The other issue I worried about with the hashing was if we wanted to change the content of the message (e.g. adding a new field, or changing the shape of the data structure, or even editing a message) but it's still basically the same entity, just a new value for it. Don't worry though you can explain that use-case in real life sometime :)

**Comment by mattwynne** [comment]
Oh! How about if the server responds with the ID/hash of the message after you've posted one? Or is it actually better if the server doesn't deliberately reveal how it addresses them?

**Comment by blaine** [comment]
> The other issue I worried about with the hashing was if we wanted to change the content of the message (e.g. adding a new field, or changing the shape of the data structure, or even editing a message) but it's still basically the same entity, just a new value for it. Don't worry though you can explain that use-case in real life sometime :)

Yeah, my thought is that once messages are posted, they're immutable – edits (if we get there!) would include a pointer to the superseded message.

Divided attention right now, but as a reminder to myself, not a full explanation: the client will always receive all the messages as long as it can't say which messages it has (e.g., incorrect algorithm). We should probably add a parameter that specifies which hash algo the server is using (or switch to multihashes)

**Pull request closed without merging**